### DISCUSSION TITLE
Change private dataset behavior for unauthenticated users?

### DISCUSSION THREAD

**Discussion by pdekraker-epa** [ckan/ckan#6641] [Q&A] [answered]
I'm wondering if there is a straightforward way to change the behavior when a user who is not logged in follows a link to a private dataset or resource. Currently CKAN returns a 404 error. I would be interested in redirecting the user to the login page. Alternatively, having a response other than "Not Found" might be more helpful.

**Comment by Zharktas** [comment, marked as answer]
It's a common security thing to return 404 instead of 403 for private objects. Otherwise someone could crawl through the URLs and see which datasets exist and perhaps reveal confidential information. But if you really want to change the behavior, you need to modify your CKAN for it, there aren't an interface for it.

**Reply by ThrawnCA** [reply]
We built a solution for this a while ago in one of our plugins, https://github.com/qld-gov-au/ckanext-qgov/blob/68e8dcdc2e8d23b42e681187eac262c6e8461427/ckanext/qgov/common/views/dataset.py#L46

Would it be feasible to incorporate it into the core as an optional behaviour controlled by a `ckan.auth` config value?

**Comment by ThrawnCA** [comment]
I have opened https://github.com/ckan/ckan/pull/7036 to make the behaviour configurable.

**Reply by smotornyuk** [reply]
Merged into master branch.

Setting `ckan.auth.reveal_private_datasets = True` switches CKAN into revealing-mode.
